src/cargo/core/workspace.rs

```diff
@@ -0,0 +1,421 @@
+                }
+                WorkspaceConfig::Member { root: Some(ref path_to_root) } => {
+                    let path = manifest_path.parent().unwrap()
+                                            .join(path_to_root)
+                                            .join("Cargo.toml");
+                    debug!("find_root - pointer {}", path.display());
+                    return Ok(Some(paths::normalize_path(&path)))
+                }
+                WorkspaceConfig::Member { root: None } => {}
+            }
+        }
+
+        let mut cur = manifest_path.parent().and_then(|p| p.parent());
+        while let Some(path) = cur {
+            let manifest = path.join("Cargo.toml");
+            debug!("find_root - trying {}", manifest.display());
+            if let Ok(pkg) = self.packages.load(&manifest) {
+                match *pkg.workspace_config() {
+                    WorkspaceConfig::Root { .. } => {
+                        debug!("find_root - found");
+                        return Ok(Some(manifest))
+                    }
+                    WorkspaceConfig::Member { .. } => {}
+                }
+            }
+            cur = path.parent();
+        }
+
+        Ok(None)
+    }
+
+    /// After the root of a workspace has been located, probes for all members
+    /// of a workspace.
+    ///
+    /// If the `workspace.members` configuration is present, then this just
+    /// verifies that those are all valid packages to point to. Otherwise, this
+    /// will transitively follow all `path` dependencies looking for members of
+    /// the workspace.
+    fn find_members(&mut self) -> CargoResult<()> {
+        let root_manifest = match self.root_manifest {
+            Some(ref path) => path.clone(),
+            None => {
+                debug!("find_members - only me as a member");
+                self.members.push(self.current_manifest.clone());
+                return Ok(())
+            }
+        };
+        let members = {
+            let root = try!(self.packages.load(&root_manifest));
+            match *root.workspace_config() {
+                WorkspaceConfig::Root { ref members } => members.clone(),
+                _ => bail!("root of a workspace inferred but wasn't a root: {}",
+                           root_manifest.display()),
+            }
+        };
+
+        match members {
+            Some(list) => {
+                let root = root_manifest.parent().unwrap();
+                self.members.push(root_manifest.clone());
+                for path in list {
+                    let manifest_path = root.join(path).join("Cargo.toml");
+                    try!(self.packages.load(&manifest_path));
+                    debug!("find_members - explicit {}", manifest_path.display());
+                    self.members.push(manifest_path);
+                }
+                Ok(())
+            }
+
+            None => self.find_path_deps(&root_manifest),
+        }
```

Comment:
Ah this is actually an implementation strictly following the RFC. I have come to think that it may be better to just always transitively follow members of the workspace via path dependencies to find more members, however, as that's somewhat more intuitive and more ergonomic I think.